File: src/main/java/twopointers/ThreeSum.java
package twopointers;
import java.util.*;

public class ThreeSum {

    // O(NLOGN
    // Solution: This is like a combination of two sum II and Two Sum on a sorted array.
    // Go through the entire array and for each value, consider it as the first number being added.
    // Then, use the two pointers technique in sorted two sum to close in on our target.
    // The array needs to be sorted for this to work.
    // Also, to filter out duplicate values we need to skip over them in the array when we see them, much like in subsets ii.
    // The tricky part here is that this has to be done in both the outer and inner loop.
    public static List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums); // in place sort algo.

        List<List<Integer>> toRet = new ArrayList<>();

        for (int i = 0; i < nums.length; i++) {
            int l = i + 1, r = nums.length - 1;


            while (l < r) { // stop trying if r is greater than nums.length or if l goes to i
                int sum = nums[l] + nums[r] + nums[i];

                if (sum == 0) {
                    toRet.add(List.of(nums[i], nums[l++], nums[r--]));
                    while (l < r && nums[l] == nums[l - 1]) l++;
                } else if (sum < 0) {
                    l++;
                } else {
                    r--;
                }
            }

            // skip over duplicates...
            while (i + 1 < nums.length && nums[i + 1] == nums[i]) i++;
        }

        return toRet;
    }


    public static void main(String[] args) {
        int[] in = new int[] {0, 0, 0};
        List<List<Integer>> output = threeSum(in);

        System.out.printf("Input: %s\nFinal Result: %s", List.of(in), output);
    }
}

File: cpp/twopointers/threeSum.cpp
#include <iostream>
#include <vector> 
using namespace std;

class Solution {
  public:
    vector<vector<int>> threeSum(vector<int>& nums) {

      vector<vector<int>> toRet;
      // Step one: Sort the input array. For this algorithm to work our input must be sorted.
      sort(nums.begin(), nums.end());



      // For the sorted array, we need to iterate from left to right.
      // 'i' will serve as the anchor point.
      for (int i = 0; i < nums.size(); i++) {

        int l = i + 1;
        int r = nums.size() - 1;

        while (l < r) {
          int sum = nums[i] + nums[l] + nums[r];
          if (sum == 0) {
            toRet.push_back({nums[i], nums[l], nums[r]});
            // in order to prevent duplicates, keep iterating while l is the same as the current element.
            while (l + 1 < nums.size() && nums[l + 1] == nums[l] ) {
              l++;
            }

            r--;
            l++;
          }
          else if (sum < 0) {
            l++;
          } else {
            r--;
          }
        }

        while (i + 1 < nums.size() && nums[i] == nums[i + 1]) i++; 
      }

      return toRet;
    }
};


int main() {
  Solution s;
  vector<int> input = {-1, 0, 1, 2, -1, -4};
  cout << "For input: [";
  copy(input.begin(), input.end(), ostream_iterator<int>(cout, " " ));
  cout << "]" << endl;

  vector<vector<int>> res = s.threeSum(input);

  cout << "Solutions: [";

  for (vector<int> vec : res) { 
    cout << "[";
    copy(vec.begin(), vec.end(), ostream_iterator<int>(cout, " "));

    cout << "]" << endl;
  }
  cout << "]";
}
